include/bluetooth/services/ots.h

```diff
@@ -708,6 +780,230 @@ int bt_ots_init(struct bt_ots *ots, struct bt_ots_init *ots_init);
+			       struct bt_ots_client *otc_inst);
+
+/** @brief Select the last object.
+ *
+ *  @param conn         Pointer to the connection object.
+ *  @param otc_inst     Pointer to the OTC instance.
+ *
+ *  @return int         0 if success, ERRNO on failure.
+ */
+int bt_ots_client_select_last(struct bt_conn *conn,
+			      struct bt_ots_client *otc_inst);
+
+/** @brief Select the next object.
+ *
+ *  @param conn         Pointer to the connection object.
+ *  @param otc_inst     Pointer to the OTC instance.
+ *
+ *  @return int         0 if success, ERRNO on failure.
+ */
+int bt_ots_client_select_next(struct bt_conn *conn,
+			      struct bt_ots_client *otc_inst);
+
+/** @brief Select the previous object.
+ *
+ *  @param conn         Pointer to the connection object.
+ *  @param otc_inst     Pointer to the OTC instance.
+ *
+ *  @return int         0 if success, ERRNO on failure.
+ */
+int bt_ots_client_select_prev(struct bt_conn *conn,
+			      struct bt_ots_client *otc_inst);
+
+/** @brief Read the metadata of the current object.
+ *
+ *  @param conn         Pointer to the connection object.
+ *  @param otc_inst     Pointer to the OTC instance.
+ *  @param metadata     Bitfield (`BT_OTS_METADATA_REQ_*`) of the metadata
+ *                      to read.
+ *
+ *  @return int         0 if success, ERRNO on failure.
+ */
+int bt_ots_client_read_object_metadata(struct bt_conn *conn,
+				       struct bt_ots_client *otc_inst,
+				       uint8_t metadata);
+
+/** @brief Read the data of the current selected object.
+ *
+ *  This will trigger an OACP read operation for the current size of the object
+ *  with a 0 offset and then expect receiving the content via the L2CAP CoC.
+ *
+ *  @param conn         Pointer to the connection object.
+ *  @param otc_inst     Pointer to the OTC instance.
+ *
+ *  @return int         0 if success, ERRNO on failure.
+ */
+int bt_ots_client_read_object_data(struct bt_conn *conn,
+				   struct bt_ots_client *otc_inst);
+
+/** @brief Directory listing object metadata callback
+ *
+ * If a directory listing is decoded using bt_ots_client_decode_dirlisting(),
+ * this callback will be called for each object in the directory listing.
+ *
+ *  @param meta The metadata of the decoded object
+ *
+ *  @return int   BT_OTS_STOP or BT_OTS_CONTINUE. BT_OTS_STOP can be used to
+ *                stop the decoding.
+ */
+typedef int (*bt_ots_client_dirlisting_cb)(struct bt_ots_obj_metadata *meta);
+
+/** @brief Decode Directory Listing object into object metadata.
+ *
+ *  If the Directory Listing object contains multiple objects, then the
+ *  callback will be called for each of them.
+ *
+ *  @param data        The data received for the directory listing object.
+ *  @param length      Length of the data.
+ *  @param cb          The callback that will be called for each object.
+ */
+int bt_ots_client_decode_dirlisting(uint8_t *data, uint16_t length,
+				    bt_ots_client_dirlisting_cb cb);
```

Comment:
Can we remove these APIs and make decoding automatic inside the implementation? The metadata of each object can be reported with _obj_metadata_read_ callback.

Reply:
That is a possibility, although there is a difference between reading the directory object and the metadata of other object.
Propose to postpone for now: https://github.com/zephyrproject-rtos/zephyr/issues/42819